### PR TITLE
Initial support for Nullable Reference Types.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet-bot Test prtest/win/dbg/unit32 please.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: docs/features/NullableReferenceTypes/Nullable reference types.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+This approach is attractive, but is difficult to add to a language where every type has a default value. For instance, a newly created array will contain all nulls. Also, such systems are notorious for problems around initialization and cyclic data structures, unless non-null types are permitted to at least temporarily contain null.
+
+On top of these issues come challenges stemming from the fact that C# already has null-unsafe types, that allow both null values and dereferencing. How can a safer approach be added to the language without breaking changes, without leaving the language more complex than necessary and with a natural experience for hardening existing code against null errors in your own time?
+
+# Approach: Nullable reference types plus nullability warnings
+
+The approach suggested here consists of a couple of elements:
+
+* Add a notion of safe nullable reference types `T?` to the language, in addition to the current ones `T` which we will now call non-nullable.
+* Detect and warn on cases where nullable types are dereferenced, or when null is values are assigned to (or left in) non-nullable variables.
```
Comment (bkoelman) [inline docs/features/NullableReferenceTypes/Nullable reference types.md]: Something not right in this sentence. Should it be "null-ish" maybe?

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs [hunk 2]
```diff
@@ -1523,49 +1933,492 @@ public override BoundNode VisitLocalDeclaration(BoundLocalDeclaration node)
...
+            if (_performStaticNullChecks && this.State.Reachable)
+            {
+                // TODO: Try to infer based on the operator
+                this.State.ResultIsNotNull = null;
+            }
+
+            return null;
+        }
+
+        protected override void AfterNestedBinaryOperatorHaveBeenVisited(BoundBinaryOperator binary)
```
Comment (bkoelman) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: Operator**Has**BeenVisited?

File: docs/features/NullableReferenceTypes/Nullable reference types.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
+# Proposal: Static null checking in C#
```
Comment (khellang) [inline docs/features/NullableReferenceTypes/Nullable reference types.md]: This doesn't render very well; `Proposal: Static null checking in C`. I think the `#` in `C#` needs to be escaped.

Comment (AlekseyTs) [inline docs/features/NullableReferenceTypes/Nullable reference types.md]: @dotnet-bot Test prtest/win/dbg/eta please.

File: docs/features/NullableReferenceTypes/Nullable reference types.md (hunk 1 above)
Comment (AlekseyTs) [inline docs/features/NullableReferenceTypes/Nullable reference types.md]: CC @MadsTorgersen

**Comment by davkean** [comment]
Hmm, the ETA failure could be caused by the open -> internal pump, @amcasey, @agocke? 

http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_dbg_eta/2038/

retest prtest/win/dbg/eta please

**Comment by davkean** [comment]
Closing/reopening to pick up latest

**Comment by dpoeschl** [comment]
retest prtest/win/dbg/eta please // Still complained about the pump not running on the retest yesterday. Trying again.

**Comment by dpoeschl** [comment]
retest prtest/win/dbg/eta please

**Comment by davkean** [comment]
This is to a feature branch, which doesn't exist on the internal side, so I'm guessing this expected? tag @jasonmalinowski @mmitche @agocke whom have some git foo, to tell if this expected or not.

**Comment by jasonmalinowski** [comment]
Looking at the log, yeah, it looks like it tries finding an internal branch with the same name as the public branch.

**Comment by AlekseyTs** [comment]
I am going to merge this PR, will respond to any additional feedback later.

**Pull request merged**